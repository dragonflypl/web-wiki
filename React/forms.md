# Forms in React

## Resources

- [Why is Formik so cute?](https://medium.com/@hygorzorak/why-is-formik-so-cute-7b32800ef878)
- [Should you store your form state in Redux?](https://goshakkk.name/should-i-put-form-state-into-redux/)

## TODO

- Check how <https://github.com/jquense/yup> integrates with `Formic`
