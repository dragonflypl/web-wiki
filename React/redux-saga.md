## What is Redux Saga

* It a process (saga) manager - it starts/stops sagas
* Redux middleware for managing side-effects (async communication: db, timers, api calls) i.e. anything that does not fit into reducers / selectors
* More sophisticated than `redux-thunk`
* it is black box that consumes and emits actions or side-effects
* effective for **apps with async operations**
* maintains continously running processes called _sagas_
* it uses effects and plain actions to coordinate sagas

## What is a Saga

Saga is a long running "thread" / function.

Saga is ES6 Generator function that yields effects.

* Listens for actions and dispatches actions (using effects)
* It is long running background process
* Creates application's side effects
* Redux-saga is process manager, that sagas can communicate with

Redux-saga works with generator functions and yield keyword & Promises.

## Effects

Effect objects cause side-effects when yielded **inside a Redux Saga app**.

Effect is utility of redux saga.

Effects are interpreted by Redux Saga.

Sagas yield effects e.g. `yield effects.take('SOME_ACTION');`

### take / put

`take` Pauses saga execution until action is dispatched:

```js
const actionPayload = yield effects.take('SOME_ACTION');
```

`put` dispatches an action.

So put & take can be used to communicate between sagas.

### call / apply

Instead of calling some function directly, `call` effect will do it.

Apply? Don't get it...

### fork

Just like call, except values yielded from forked function cannot be captured.

And also fork does not pause code execution: caller continues without pausing.

Forks run independently.

### takeEvery

Combination of take and fork. For each action it forks new method.

It returns immadetely (contrary to take).

`takeEvery('SOME_ACTION', methodToFork)`

### cancel & cancelled

These are two effects for cancelling the process that was created with fork, and for checkking cancellation status.

### takeLatest

Combination of fork, cancel, takeEvery.

Each time action arrives, previous process is cancelled before new one is forked.

### spawn

Just like fork, however spawned process is not a child of a caller process.

If caller cancells or throws error, then spawned process won't be affected.

### all

This effect takes array of take effects and waits until all resolve and then caller is resumed.

## Channels

### Action Channel

Actions channel works like a buffer (a queue) that stores all actions that were dispatched. All actions get executed, nothing is lost.

Compared to regular `take('SOME_ACTION')`, if process is doing something (like delay of any other async operation) and waiting, then subsequent actions will be lost (won't be processed).

If we `take(SOME_CHANNEL)`, then it will take oldest action (`FIFO`).

### Event Channel

Wraps and outside source of evenst (i.e. WebSockets).

Event channels convert some event into actions that can work with `take` and other effects.

### Generic Channel

Establishes communication between two channels.

## Testing

There're two ways of testing:

### Unit testing (testing effects)

This method does not require store nor application state.

However it is stricly connected to implementation details. So even if the output of saga is the same, but implementation deails has changed, then test would need to be rewriten.

### E2E tests

Requires more mocking (like store, state, API's etc).

On the bright side it focuses or how application looks like before and after the saga.

## Installation

Simple as :

```js
import createSagaMiddleware from 'redux-saga';
const sagaMiddleware = createSagaMiddleware();
// and use this middleware
```

## Redux-saga and Promises

They play nicely: library is automatically waiting for the promise to resolve before advancing to next step.

## Resources

* Sample app:
  * https://github.com/danielstern/redux-saga-cart
  * https://github.com/danielstern/redux-saga-shopping-cart-server
