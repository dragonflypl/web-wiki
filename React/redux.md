# Redux

## Redux & React - how to connect

react-redux connects React container components with Redux store. 

- `connect` function creates container components and 

```
export default connect(mapStateToProps, mapDispatchToProps)(SomeComponent);
```

mapStateToProps defines what parts of store should be exposed to component. It's first argument is `store`, and second is `ownProps` (properties passed to generated container components):

`mapStateToProps = (state, ownProps) => { return { something: ownProps.sth === state.sth } }`

`mapDispatchToProps` exposes actions that component can execute. There's three ways to define actions, most common is to use `bindActionCreators`:

`function mapDispatchToProps(dispatch, ownProps) { return { actions: bindActionsCreators(actions, dispatch) } }`

and in component use `this.props.actions.loadCourses()`

Container component will make state & dispatch to props + it will add 

- `Provider` components attaches application to Redux store. It provides redux store to all components automatically (via react context)

### Reselect library

Use is in mapStateToProps for caching, as reselect enables memoization: https://github.com/reactjs/reselect

## Folder structure

- store actions in `actions` folder. For each logical part, create separate actions file e.g. `courseActions.js`

This is action creator, type property is mandatory:

`export function createCourse(course) { return { type:' CREATE_COURSE', course }}`

- store reducers in `reducers` folder. 

This is a reducer (e.g. `courseReducer.js`) :

`export default function courseReducer(state = [], action) { switch(action.type) ... }`

Create a root reducer in `index.js`:

```
import {combineReducers} from 'redux'
import courses from './courseReducer'; // some reducer

const rootReducer = combineReducers({courses})

export default rootReducer;
```

Use `combineReducers`  for reducers composition.

## Three principles

- single immutable state tree (object) / store (it stores data & ui state) contains everything in the application that can change (data & UI state)
- state tree is readonly. actions trigger changes. Action is minimal representation of change to data. Structure must have `type` property. Components just dispatch actions and their work is done.
- reducers update store: state mutations are described by pure functions (reducers) & action

### Enforcing immutability

Use `redux-immutable-state-invariant`. Use is only in development.

or 

Use `Immutable.js`

or 

`deep-freeze`.

## Reducers

Functions that take current state and actions and return new state.

Each reducer is responsible for its slice of the state.

Reducers are pure functions!

A single action can be handled by all, some or none of reducers.

Reducer should initialize state if previous one is undefined:

```
const counter = (state = 0, action) => {
	case...
	default:
		return state;
}
```

## Actions & Action creators

Action creators are functions that create actions, e.g.

```
rateCourse(rating) { return { type: 'RATE_COURSE', rating } }
```

## Store API

- createStore

```
const store = createStore(rootReducer);
```

- store.dispatch(action)
- store.subscribe(listener)
- store.getState()
- replaceReducer(nextReducer)
