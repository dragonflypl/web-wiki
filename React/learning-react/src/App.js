import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';

import VisibleTodoList from './components/VisibleTodoList';
import AddTodo from './components/AddTodo';
import Footer from './components/Footer';

class App extends Component {

  render() {
    return (
      <div>
        <div className="App">
          <header className="App-header">
            <img src={logo} className="App-logo" alt="logo" />
            <h1 className="App-title">Welcome to React</h1>
          </header>
          <hr />
          <AddTodo />
          <VisibleTodoList />
          <Footer />
        </div>
      </div >
    );
  }
}

export default App;
