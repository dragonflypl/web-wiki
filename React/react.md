
## React app

First decision to make is to decide on components structure. 

## React Components

Component needs to be mount in to the browser. `ReactDOM.render(component, placeToMountTheComponent)`.

Naming convention is to use title case for naming the components e.g. `Button`.

React component returns a description of user interface in the form of React element (`render` method or component function). 
Rendering of this React elements is done via `ReactDOM` library. 

### Container components

A components that does anything beside presenting UI is known as a container component `ArticleContainer`:

```javascript
const ArticleContainer = (props, { store }) => {
  return <Article {...props} store={store} />;
};
```

It is commonly used to access e.g. context and pass it down to presentational component (`Article`) as props.

- It is aware of Redux.
- Focus on how things work.
- subscribe to redux state
- dispatch redux actions
- generate by react-redux

### Presentational components

- Focus on how things look
- unaware of redux
- read data from props
- invoke callbacks on props
- written by hand, often only render function

### Pure components

`React.PureComponent` will rerender only if shallow comparison of props & state detect changes.

To force updates, call `forceUpdate` method.

### Higher-order component (function)

> Check out this ultimate guide: <https://medium.freecodecamp.org/higher-order-components-the-ultimate-guide-b453a68bb851>

HoC is a function that takes a components and returns new one.

It is a technique from FP (higher-order functions). Just as the higher-order function creates a new function, the HoC creates a new component.

HoC is a  **function**  that  **accepts**  **a component**  and  **returns a new component that renders the passed one**.  This new component is enhanced with an additional functionality.

> const HoC = BaseComponent => EnhancedComponent

`recompose` is a library that enables HoC composition and provides dozens of generic HoC.

 (usually this is a generic function that generates a container components from presntational components):

```javascript
import React from 'react';
import PropTypes from 'prop-types';

const storeProvider = (extraProps) => (Component) => {
  return class extends React.Component {
    static displayName = `${Component.name}Container`;
    static contextTypes = {
      store: PropTypes.object,
    };

    render() {
      return <Component
        {...this.props}
        {...extraProps(this.context.store, this.props)}
        store={this.context.store} />;
    }
  };
};

export default storeProvider;
```

And usage:

```javascript
function extraProps(store, originalProps) {
  return {
    author: store.lookupAuthor(originalProps.article.authorId),
  };
}

export default storeProvider(extraProps)(Article);
```

### Component keys

If components are rendered from arrays, add `key` property to component. React will use it to identify components and optimize rendering: `<SomeComponent key={data.id} />`.

### (stateless) Function Component

Simplest form of component: function that takes inputs (properties, props) and returns JSX.

```javascript
const Button = () => {
  return (
    <button>Go</button>
  )
}
```

To mount this component: `ReactDOM.render(<Button />, mountNode)`.

To use props, add argument and use props in JSX using `{}`

```javascript
const Button = (props) => {
  return (
    <button>{props.label}</button>
  )
}
```

and `ReactDOM.render(<Button label="Go" />, mountNode)`

### Class Component

> Use them only when you have a state to manage or personalized event handlers or need lifecyclehooks or access to DOM is needed or may child functions (for performance reasons, so that they're on prototype. In function components child functions would be created for each component thus bloating the memory)

Apart from props, it can have private internal state. State can be changed, props not! Class component can change only its internal state.

State is available via `this.state`. State is available only to component it belongs, it is private.

```javascript
class Button extends React.Component {

  // not mandatory
  constructor(props) {
  	super(props);
  }
  
  state = { counter: 0 };

	handleClick = () => {
  	//this.setState({ counter: this.state.counter + 1 });
    
    // or safer method - updater function: use it if you access previous state value
    // The output of the updater is shallowly merged with prevState.
    this.setState((prevState) => ({ 
    	counter: prevState.counter + 1
    }));
  }
  
	render() {
  	return (
    	<button onClick={this.handleClick}>{this.props.label} {this.state.counter}</button>
    )
  }
}

ReactDOM.render(<Button label="Counter: " />, mountNode)
```

### Styling components

Inline styles can be used: `style={{ display: 'inline-block'}}` as well as classes (`className` attribute) e.g.

`<button className="btn" style={{color: 'blue'}} onClick={this.handleClick}>{this.props.label} {this.state.counter}</button>`

#### Theming

react-toolbox is great example of how theming can be done with css modules.

### Access to the DOM elements

In order to get access to element in JSX, use `ref` attribute: `<input ref={(input) => this.someInput = input} />`. This `ref` function will execute when element is mounted.

```
const Item = (props) => {
	return (
  	<div ref={(el) => console.log('Logged output', el.innerHTML)}>{props.data}</div>
  )
}
```

Alternative to `ref` is controlled components:

- use `value="this.state.field"` to bind state to input
- use `onChange="this.setState({ field: event.target.value })" to update the value

```javascript
class Form extends React.Component {
	state = { inputValue: 'Default' };
	render() {
  	return (
    	<div>
        <input value={this.state.inputValue} onChange={(event) => this.setState({ inputValue: event.target.value })}  />      
        {this.state.inputValue}
      </div>
    )
  }
}
```

### Events

- `onSubmit` on form, remember to call `event.prefentDefault()`
- `onClick` on any

Any event handler receives `event` argument that is wrapper around native even object.

### Default properties

In order to provide default values for component properties, use `static defaultProps = { ... }` inside component.

### Rendering children

Component props have `children` property that has access to child components.  This is similar to transpilation in Angular.

## JSX

React compiles JSX into JavaScript with `React.createElement` etc. (JavaScript representation of the DOM). Both syntax can be used, yet JSX is simpler.

### Loops

In JSX we can use JavaScript, so rendering component for each element in array is as simple as mapping to React Component. Also react supporst object spread operator that passess all object properties to component:

```javascript
const Item = (props) => {
	return (
  	<div>{props.data}</div>
  )
}

const Items = (props) => {
	return (
  	<div>{props.data.map(x => <Item {...x} />)}</div>
  )
}

class App extends React.Component {

	state = { data: [
    { data: 'One' },
    { data: 'Two' },
    { data: 'Three' }
  ] }

	render() {
  	return (
    	 <Items data={this.state.data} />
    )
  }
}

ReactDOM.render(<App />, mountNode)
```

## Virtual DOM

Using JavaScript to render HTML allows React to have a virtual representation of HTML in-memory (Virtual DOM).

React uses this concept to render and HTML tree virtually first (in-memory). When state changes and new tree is created, instead of writing whole tree, React renders only difference. This is called tree reconciliation

- What is tree reconciliation: it is a process of comparing old tree with new tree and updating only the difference in the DOM

## Ajax

Use case is simple. Just call `setState` in callback and react will update. 

## Server side rendering

Use: `import ReactDOMServer from 'react-dom/server'`

It:

- enables SEO, whole application in rendered on server.
- faster rendering on clients that are slow

When doing server side rendering with initial data, render this data to the client as well (e.g. by rendering it to `index.html` and make it a global variable), so that initial client side rendering renders exactly the same content (so it does not rerender the application).

## Context API

Context API enables sharing global stuff, without passing them as props down & down & down to child components.

Just define `getChildContext` & `childContextTypes` in parent + `contextTypes` in child, and second argument of components constructor will be context (for functional components) or `this.context` (for class components):

```javascript
function SomeComponent(props, context) {}
```

Using Context API explicitly is tricky, and at least makes testing harder (global dependency). To address some of the issues while testing, use Enzyme that does shalow rendering (react-test-renderer is tree renderer).

Don't use it, unless you're using some kind of provider component along with HoC that binds context with base component's props.

## Type checking

It's possible to guard against invalid properties passed to components with `PropTypes`:

```javascript
Article.propTypes = {
  article: PropTypes.shape({
    title: PropTypes.string.isRequired,
    body: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
  })
};
```

Another solution is to use `Flow` - typechecker: <https://flow.org/>

## Testing

- `react-test-renderer` - for snapshots
- `enzyme` - for interaction. `shallow` does not render children. To test lifecycle methods and children interaction, use `mount`. 
- `react-addons-test-utils` - used by enzyme

### Types of testing

- Unit testing - enzyme's shallow can be useful
- Interaction testing - use enzyme (e.g. clicking etc)
- Structural testing - possible with snapshots, usually this is what stateless functional components need
- Style testing (like PhantomCSS)

### Enzyme & shallow rendering

Enzyme enables component creation and nice syntax for querying component's content etc... Also it enables shalow rendering (only component being created is rendered, child components not), this is perfect for unit testing.

Instead of `renderer.create`, use:

```javascript
import { shallow } from 'enzyme';
const wrapper = shallow(<ArticleList {...testProps} />);
wrapper.find('a').simulate('click'); // uses JSDOM under the hood
wrapper.instance(); // to access instance,methods etc...
```

### Snapshot testing

Renderer can be used to do it;

```javascript
import renderer from 'react-test-renderer'
const tree = renderer.create(<div>Hello</div>).toJSON()
expect(tree).toMatchSnapshot();
```

You can use enzyme with this rendered to test shapshots of only parts of components.

# Tools

- `jest`
- https://jscomplete.com/repl/ : playground
- react-devtools
- https://github.com/axios/axios
- https://github.com/facebook/create-react-app : create react app
- `eslint`, `eslint-plugin-react`, `babel-eslint` along with <https://github.com/samerbuna/.files/blob/master/.eslintrc.js>
- `nodemon` with script `nodemon start {filename} --watch --interpreter babel-node`
- `babel-cli` along with presets: `react`, `env`, `stage-2` (`babel-preset-react`, `babel-preset-env`, `babel-preset-stage-2`)

# Creating reusable components

## Styling

### React Inline styles

You can write styles as javascript objects. However, even though it has some benefis (styles can be dynamic, deterministic), it does not support more advances features (autoprefixing, animations, media queries, pseudoclasses etc...)

### Naming Scheme (e.g. BEM)

Use BEM for creating class names that won't collide with / leak out from comonents.

### CSS Modules

Use CSS Modules - they work like BEM out of the box. Just enable modules in `css-loader`

Now you can do really short class names and be safe that no collision will happen.

If you enable modules, you need to use it in whole application.

### CSS in JS

I'm interested in it. If my mind changes check `styled-components`.

## Architecture

Atoms -> Molecules -> Ogranisms

### Atom

Button, Label etc. - this is basic building block.

Each component code (js/styles/tests) in it's folder inside `components` folder (this is used by almost all component libraries) e.g.:

`https://github.com/react-toolbox/react-toolbox/tree/dev/components/dropdown`

Hints:

- when wrapping elements, honor native API (make props have the same names as native DOM elements like className, value, maxLength)
- for arrays, use plurals (users), not suffixes (userList)
- use spread operator (with desctucturing `{ name, ...rest}` ) to pass props `{...props}` when possible
- create formatting components! (`<Cash>6</Cash>` -> `$6`)

### Organisms

Make them dumb, do not create mini apps.

Organisms still should not make any API calls. They still should receive data via props.

Otherwise they can lead to overfetching and displaying out of sync data and other problems.

## Environment

### General boilerplate

- create-react-app : application oriented

```
npm i -g create-react-app
# will create app in current directory
create-react-app ./
# for ejecting
npm run eject
```

- nwb : single component oriented
- list of other boilerplates: https://www.javascriptstuff.com/react-starter-projects/

### Documentation tool

- react storybook
- react styleguide
- react-styleguide-generator
- bluekit

### Custom documentation app

With `react-docgen`.

# Resources

- <https://github.com/ReactTraining> : interesting repos
- http://sandny.com/2017/10/30/babel-express-js-node-js-nodemon-to-build-a-node-js-server-with-hot-reloading/ - babel with nodemon and react
- slack.jscomplete.com - slack of author of React courses
- https://app.pluralsight.com/library/courses/react-js-getting-started/table-of-contents
- https://app.pluralsight.com/library/courses/reactjs-advanced/table-of-contents 
  - https://github.com/jscomplete/advanced-react

# FAQ

- what is componentDidMount
- what is:
  - smart
  - dumb / presentational / stateless component: only handles UI, does not deal with state. State should be handled by higher level container components. They don't have access to local state.

# TODO

- read https://cdb.reacttraining.com/react-inline-functions-and-performance-bdff784f5578
- do last two modules: https://app.pluralsight.com/library/courses/reactjs-advanced/table-of-contents
- https://app.pluralsight.com/library/courses/react-redux-react-router-es6/table-of-contents
  - do second module: Env Setup
