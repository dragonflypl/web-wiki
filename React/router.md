# React router

`npm i react-router-dom`

```javascript
import { BrowserRouter as Router, Route } from "react-router-dom";

export default ({ store }) => {
  return (<Provider store={store}>
    <Router>
      <Route path='/:filter?' component={App} />
    </Router>
  </Provider>);
}
```

## Route params

`Route` component specifies component to be rendered when `path` is matched.

Each routed component (e.g. `App`) receives `match` property from the router:

`<VisibleTodoList visibilityFilter={this.props.match.params.filter} />`

Another way to access route params is HoC `import { withRouter } from 'react-router'`. This HoC will inject `match` as the property to wrapped component:

```javascript
// now ownProps contains match property
const mapStateToProps = ({ todos }, { match }) => {
  const visibilityFilter = match.params.filter || 'all';
  return {
    todos: todos.filter(x => {
      return (visibilityFilter === 'all') ||
        (visibilityFilter === 'completed' && x.completed) ||
        (visibilityFilter === 'active' && !x.completed);
    })
  }
}
```

### Params

`:paramName` is the syntax for specifing route parameters.

### Optional params

`<Route path='/:filter?' component={App} />` - add `?` after param name to make it optional.

## Navigation links

Simple as that, use `NavLink`:

```javascript
import React from 'react';
import { NavLink } from 'react-router-dom';

export default function FilterLink({ filter, ...rest }) {
  return <NavLink activeStyle={{
    textDecoration: 'none',
    color: 'black'
  }} to={'/' + filter} {...rest} />
}
```
