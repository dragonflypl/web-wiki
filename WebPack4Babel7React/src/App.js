import React from 'react';
import style from './App.css';

const Hello = () => (
  <div>
    Hello <span className={style.name}>React!</span>
  </div>
);

const App = () => {
  return <Hello />;
};

export default App;
