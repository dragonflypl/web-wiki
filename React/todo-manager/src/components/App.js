import React from 'react';
import logo from './../logo.svg';
import './App.css';
import VisibleTodoList from './VisibleTodoList';
import AddTodo from './AddTodo'
import Footer from './Footer'

export default ({ match }) => {
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <h1 className="App-title">Welcome to React</h1>
      </header>
      <hr />
      <AddTodo />
      <VisibleTodoList />
      <Footer />
    </div >
  );
}
