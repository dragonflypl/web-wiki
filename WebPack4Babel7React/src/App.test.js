import React from 'react';
import TestRenderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import App from './App';

describe('App', () => {
  it('should render', () => {
    const component = TestRenderer.create(<App />);
    expect(component).toMatchSnapshot();
  });

  it('should render Hello', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find('Hello').length).toBe(1);
  });

  it('should shallow render', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });
});
