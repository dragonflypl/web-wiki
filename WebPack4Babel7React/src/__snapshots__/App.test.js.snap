// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render 1`] = `
<div>
  Hello 
  <span
    className="name"
  >
    React!
  </span>
</div>
`;

exports[`App should shallow render 1`] = `<Hello />`;
